URL Module Implementation

dinsfire64 opened the pull request:
Let me know if I missed any of your comments.

dinsfire64 committed:
Initial module working for getting titles from url

dinsfire64 committed:
Implemented checks for http/s, check for public IP range, download partial page from website

dinsfire64 committed:
cleanup

dinsfire64 committed:
change to requests module

dinsfire64 committed:
forgot to include headers in get request

dinsfire64 committed:
initial unit tests

dinsfire64 committed:
Merge https://github.com/calzoneman/bavi into htmlscraper

dinsfire64 committed:
Initial regex trigger implemented

dinsfire64 committed:
updated unit test

dinsfire64 committed:
fixed empty statement when valid url has no response

dinsfire64 committed:
removed unneeded variable in unit test

dinsfire64 committed:
attempt to get mock working with bot class

dinsfire64 committed:
Fixed url order, fixed bot mock tests, add multi url test

dinsfire64 committed:
added emoji; not IDNA standard 😒

dinsfire64 committed:
removed suggested ignoring regex

dinsfire64 committed:
Added mock tests for url

dinsfire64 commented:
Let me know if you have any more comments.

dinsfire64 committed:
added Requests module to requirements

dinsfire64 committed:
More requirements.

calzoneman commented on bavi/modules/url.py:
```diff
@@ -0,0 +1,109 @@
+
+URL_REGEX = r'(https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=\u263a-\U0001f645]{1,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*))'
+
+IP_REGEX = r"(https?:\/\/([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])(?<!172\.(16|17|18|19|20|21|22|23|24|25|26|27|28|29|30|31))(?<!127)(?<!10)(?<!0)\.([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])(?<!192\.168)(?<!172\.(16|17|18|19|20|21|22|23|24|25|26|27|28|29|30|31))\.([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])(?<!\.255$)([-a-zA-Z0-9@ ...
```
Is this `IP_REGEX` behemoth needed?  It seems like the `URL_REGEX` could be relaxed a bit to accomodate IPs.

I also note that your URL regex won't match `.technology` as a TLD, since I see `[a-z]{2,6}`.

calzoneman commented on bavi/modules/url.py:
```diff
@@ -0,0 +1,109 @@
+        valid_urls = get_valid_urls(message)
+
+        # make sure we have any good url, otherwise ignore
+        if (len(valid_urls) > 0):
```
parens

calzoneman commented on bavi/modules/url.py:
```diff
@@ -0,0 +1,109 @@
+
+def init(bot):
+    bot.add_command('title', title_command)
+    bot.add_matcher(compiled_url_regex, title_command)
```
pass `priority='low'`.

calzoneman commented on bavi/modules/url.py:
```diff
@@ -0,0 +1,109 @@
+                response = get_user_readable_response(url)
+
+                if len(response) > 0:
+                    bot.say(target, response)
```
Rate limiting responses is probably something the core should be responsible for, but I also think you may want to sanity check the number of URLs before spitting out 10 titles at once.

calzoneman commented on bavi/modules/url.py:
```diff
@@ -0,0 +1,109 @@
+    response = ''
+
+    if (len(url) > 0):
+        #now start a request for the page title
```
Don't need this comment, the code is written cleanly so I can read what it's doing.

calzoneman commented on bavi/modules/url.py:
```diff
@@ -0,0 +1,109 @@
+
+        #TODO: bring these parameters to user defined setting.
+        #only request first 1024 bytes from server and timeout at 500ms
+        headers = {"Range": "bytes=0-1024"}
```
1. Server is not guaranteed to respond with the range
  2. Server might respond with `416 Range Not Satisfiable`
  3. Stupid websites put kilobytes of javascript crap before the title tag

Might be more worthwhile to use the streaming mode of `requests` and just take the first X kilobytes of the response (I did 1MB for sopel, but that might be overkill).

calzoneman commented on bavi/modules/url.py:
```diff
@@ -0,0 +1,109 @@
+        #TODO: bring these parameters to user defined setting.
+        #only request first 1024 bytes from server and timeout at 500ms
+        headers = {"Range": "bytes=0-1024"}
+        r = requests.get(url, headers=headers, timeout=0.5)
```
500ms seems a little low.  Somewhere around 5 seconds feels like the sweet spot between waiting for slow sites to load and not dumping no-longer-relevant URL titles into chat.

calzoneman commented on bavi/modules/url.py:
```diff
@@ -0,0 +1,109 @@
+        if r.headers['content-type'].startswith('text/html'):
+            page = UnicodeDammit(r.content, is_html=True)
+            parser = html.HTMLParser(encoding=page.original_encoding)
+            root = html.document_fromstring(r.content, parser=parser)
```
Is this just using bs4 for the Unicode conversion and then using `lxml` directly for parsing?  Doesn't bs4 already provide a convenience for this kind of thing?

calzoneman commented on bavi/modules/url.py:
```diff
@@ -0,0 +1,109 @@
+
+            title = root.find('.//title').text_content()
+
+    except Exception as e:
```
The command dispatcher will catch the error, print a command error in the channel, and log the stack trace.  I'm assuming you added this because you wanted to be able to still process other URLs even if one failed.

This snippet might be useful: https://github.com/calzoneman/bavi/blob/master/bavi/bot.py#L138

calzoneman commented on test/modules/url_test.py:
```diff
@@ -0,0 +1,184 @@
+            self.assertEqual(url_module.get_valid_urls(url), empty_list)
+
+    def test_url_full_query_stack(self):
+        for tup in example_pages:
```
`for url, title, hostname in example_pages:`

Tuple unpacking is more readable than indexing.

calzoneman reviewed (changes requested):
I think we're getting there, hopefully I won't bother you too much.

dinsfire64 commented:
Nah man you are not bothering me for these comments. I much appreciate the in depth responses that allow me to actually learn something. I'm just afraid you could have written it in a quarter of the time.

I'll work on these tonight. Thanks.

dinsfire64 committed:
Integrated calzoneman's comments
Push before moving to stream

calzoneman commented on bavi/bot.py:
```diff
@@ -8,7 +8,7 @@
 
 log = logging.getLogger('bavi')
-log.setLevel(logging.INFO)
+log.setLevel(logging.DEBUG)
```
Should default to `INFO`, we can either have a config key for this or an environment variable.

calzoneman commented on bavi/modules/url.py:
```diff
@@ -0,0 +1,108 @@
+def title_command(bot, source, target, message, **kwargs):
+    log.debug('url module triggered on: ' + message)
+
+    if len(message) > 0:
```
I think a lot of this method can be flattened because your methods implicitly return nothing if the input was blank.

For example, below you have

```
if len(thing) > 0:
  for x in thing:
```

Which is redundant.

calzoneman commented on bavi/modules/url.py:
```diff
@@ -0,0 +1,108 @@
+            log.debug('%s responded with %d', url, r.status_code)
+            
+            if r.headers['content-type'].startswith('text/html'):
+                bs = BeautifulSoup(r.content[1:512000], 'html.parser')
```
Curious why the `[1:512000]` instead of just `[:512000]`?

calzoneman commented on bavi/modules/url.py:
```diff
@@ -0,0 +1,108 @@
+    except Exception as e:
+        exception_msg = type(e).__name__ + ': ' + str(e)
+        log.exception('Could not process: %s', url)
+        self.say(event.target, exception_msg)
```
I think you meant `bot.say` because `self` is undefined here.  I would also accept just getting rid of this because from my reading of the code above, an exception in this block would be a bug in the code and not just a faulty URL, since the URL has been validated before this method.

dinsfire64 committed:
working on exception handling, needs cleanup and test cases

calzoneman commented:
Just checking in on this (no rush, I know it's Thanksgiving week) -- there's a few things we discussed offline but I think those can wait for future PRs.  I'm fine to merge if you fix the minor issues above (like `self.say`).

dinsfire64 committed:
New method of streaming title, formatting, and tests

Please read carefully. I am not 100% happy with the testing mock either as for some reason it fails with the ndash in the title of your webpage; however it works in practice.

calzoneman commented:
>Please read carefully

YOLO, I think let's just checkpoint this for now and we can fix it up as needed.  As long as there's no glaring breaks-the-entire-bot issue, I'm fine if there's an edge case that needs to be fixed later.

I am curious about your en-dash problem though.
